lib/events/athena/querier.go

```diff
@@ -0,0 +1,485 @@
+	builder strings.Builder
+	args    []string
+}
+
+// withTicks wraps string with ticks.
+// string params in athena need to be wrapped by "ticks".
+func withTicks(in string) string {
+	return fmt.Sprintf("'%s'", in)
+}
+
+func sliceWithTicks(ss []string) []string {
+	out := make([]string, 0, len(ss))
+	for _, s := range ss {
+		out = append(out, withTicks(s))
+	}
+	return out
+}
+
+func (q *queryBuilder) Append(s string, args ...string) {
+	q.builder.WriteString(s)
+	q.args = append(q.args, args...)
+}
+
+func (q *queryBuilder) String() string {
+	return q.builder.String()
+}
+
+func (q *queryBuilder) Args() []string {
+	return q.args
+}
+
+type searchParams struct {
+	fromUTC, toUTC time.Time
+	limit          int
+	order          types.EventOrder
+	startKeyset    *keyset
+	filter         searchEventsFilter
+	sessionID      string
+}
+
+func prepareQuery(params searchParams, tablename string) (string, []string) {
+	qb := &queryBuilder{}
+	qb.Append(`SELECT DISTINCT uid, event_time, event_data FROM `)
+	// tablename is validated during config validation.
+	// It can only contain characters defined by Athena, which are safe from SQL
+	// Injection.
+	// Athena does not support passing table name as query parameters.
+	qb.Append(tablename)
+	qb.Append(` WHERE event_date BETWEEN date(?) AND date(?)`, withTicks(params.fromUTC.Format(iso8601DateFormat)), withTicks(params.toUTC.Format(iso8601DateFormat)))
+	qb.Append(` AND event_time BETWEEN ? and ?`,
+		fmt.Sprintf("timestamp '%s'", params.fromUTC.Format(athenaTimestampFormat)), fmt.Sprintf("timestamp '%s'", params.toUTC.Format(athenaTimestampFormat)))
+
+	if params.sessionID != "" {
+		qb.Append(" AND session_id = ?", withTicks(params.sessionID))
+	}
+
+	if len(params.filter.eventTypes) > 0 {
+		// Athena does not support IN with single `?` and multiple parameters.
+		// Based on number of eventTypes, first query is prepared with defined
+		// number of placeholders. It's safe because we just taken len of event
+		// types to query, values of event types are passed as parameters.
+		eventsTypesInQuery := fmt.Sprintf(" AND event_type IN (%s)",
+			// Create following part: `?,?,?,?` based on len of eventTypes.
+			strings.TrimSuffix(strings.Repeat("?,", len(params.filter.eventTypes)), ","))
+		qb.Append(eventsTypesInQuery,
+			sliceWithTicks(params.filter.eventTypes)...,
+		)
+	}
+
+	if params.order == types.EventOrderAscending {
+		if params.startKeyset != nil {
+			qb.Append(` AND (event_time, uid) > (?,?)`,
+				fmt.Sprintf("timestamp '%s'", params.startKeyset.t.Format(athenaTimestampFormat)), fmt.Sprintf("'%s'", params.startKeyset.uid.String()))
+		}
+
+		qb.Append(` ORDER BY event_time ASC, uid ASC`)
+	} else {
+		if params.startKeyset != nil {
+			qb.Append(` AND (event_time, uid) < (?,?)`,
+				fmt.Sprintf("timestamp '%s'", params.startKeyset.t.Format(athenaTimestampFormat)), fmt.Sprintf("'%s'", params.startKeyset.uid.String()))
+		}
+		qb.Append(` ORDER BY event_time DESC, uid DESC`)
+	}
```

Comment:
> is there any test for pagination?

I am planning to add integration tests for whole solution and it will execute that test which checks pagination: https://github.com/gravitational/teleport/blob/f9ac4f338901c854180553328942a284fc281e96/lib/events/test/suite.go#L85-L86. However integration tests will be added in separate PR because it requires whole solution to work. 

It should not miss any events due to rounding of timestamps. I have use [following approach](https://use-the-index-luke.com/sql/partial-results/fetch-next-page#sb-row-values) and tested it locally, two events with the same timestamp are properly paginated.

Reply:
I was worried that a timestamp like `2006-01-02 15:04:05.0005` could be truncated to `2006-01-02 15:04:05.000`. If paging in descending order I think in theory you could miss some events in that scenario. But on a closer look I think all of the timestamps have millisecond precision so it shouldn't be an issue.